src/java.base/share/classes/jdk/internal/loader/ClassLoaders.java

```diff
@@ -64,19 +64,20 @@ private static void setArchivedServicesCatalog(ClassLoader loader) {
     // Creates the built-in class loaders.
     static {
         ArchivedClassLoaders archivedClassLoaders = ArchivedClassLoaders.get();
+        // -Xbootclasspath/a or -javaagent with Boot-Class-Path attribute
+        String append = VM.getSavedProperty("jdk.boot.class.path.append");
+        URLClassPath bootAppendUcp = (append != null && !append.isEmpty())
+                ? new URLClassPath(append, true)
+                : null;
         if (archivedClassLoaders != null) {
             // assert VM.getSavedProperty("jdk.boot.class.path.append") == null
             BOOT_LOADER = (BootClassLoader) archivedClassLoaders.bootLoader();
             setArchivedServicesCatalog(BOOT_LOADER);
+            BOOT_LOADER.setClassPath(bootAppendUcp);
```

Comment:
It might be clearer to set this before calling setArchivedServicesCatalog.

Reply:
Fixed. Also removed the commented assert statement.